tfhe/src/integer/server_key/radix_parallel/cmux.rs

```diff
@@ -0,0 +1,232 @@
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+    ///
+    /// // We have 4 * 2 = 8 bits of message
+    /// let size = 4;
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2_KS_PBS, size);
+    ///
+    /// let a = 128u8;
+    /// let b = 55u8;
+    ///
+    /// let mut ct_a = cks.encrypt(a);
+    /// let mut ct_b = cks.encrypt(b);
+    ///
+    /// let mut condition = sks.scalar_ge_parallelized(&ct_a, 66);
+    ///
+    /// let ct_res = sks.smart_if_then_else_parallelized(&mut condition, &mut ct_a, &mut ct_b);
+    ///
+    /// // Decrypt:
+    /// let dec: u8 = cks.decrypt(&ct_res);
+    /// assert_eq!(if a >= 66 { a } else { b }, dec);
+    /// assert_ne!(ct_a, ct_res);
+    /// assert_ne!(ct_b, ct_res);
+    /// ```
+    pub fn smart_if_then_else_parallelized(
+        &self,
+        condition: &mut RadixCiphertext,
+        true_ct: &mut RadixCiphertext,
+        false_ct: &mut RadixCiphertext,
+    ) -> RadixCiphertext {
+        let mut ct_refs = [condition, true_ct, false_ct];
+
+        ct_refs.par_iter_mut().for_each(|ct_ref| {
+            if !ct_ref.block_carries_are_empty() {
+                self.full_propagate_parallelized(ct_ref);
+            }
+        });
+
+        let [condition, true_ct, false_ct] = ct_refs;
+        self.unchecked_if_then_else_parallelized(condition, true_ct, false_ct)
+    }
+
+    pub(crate) fn unchecked_programmable_if_then_else_parallelized<F>(
+        &self,
+        condition_block: &crate::shortint::Ciphertext,
+        true_ct: &RadixCiphertext,
+        false_ct: &RadixCiphertext,
+        predicate: F,
+    ) -> RadixCiphertext
+    where
+        F: Fn(u64) -> bool + Send + Sync + Copy,
+    {
+        let inverted_predicate = |x| !predicate(x);
+
+        // Although our mul algorithm has special path for when rhs or lhs is a boolean value,
+        // we don't call it as for the ct_false we would need an extra pbs to 'invert' the
+        // ciphertext from true to false.
+        let (mut true_ct, false_ct) = rayon::join(
+            move || {
+                let mut true_ct = true_ct.clone();
+                self.zero_out_if(&mut true_ct, condition_block, inverted_predicate);
+                true_ct
+            },
+            move || {
+                let mut false_ct = false_ct.clone();
+                self.zero_out_if(&mut false_ct, condition_block, predicate);
+                false_ct
+            },
+        );
+        // If the condition was true, true_ct will have kept its value and false_ct will be 0
+        // If the condition was false, true_ct will be 0 and false_ct will have kept its value
+        true_ct
+            .blocks
+            .par_iter_mut()
+            .zip(false_ct.blocks.par_iter())
+            .for_each(|(lhs_block, rhs_block)| {
+                self.key.unchecked_add_assign(lhs_block, rhs_block);
+                self.key.message_extract_assign(lhs_block)
```

Comment:
do a bivariate PBS ? Might be more trouble than it's worth though

I like using the OR operation in that case when I had to use it for some clear bit tricks

so a bitor parallelized ?

Not sure if there may be issues with bivariate compliant PBS, otherwise this is fine

Reply:
message_extract_assign does not use a thread local lut, I see a apply_msg_identity_lut_assign which seems to call a ks_pbs or pbs_ks with pre determined table

This is also due a clean up I guess, I'll open the issue

I'm not too worried for the allocation though

Reply:
actually already have an issue : https://github.com/zama-ai/tfhe-rs-internal/issues/66

unclear what we really want and what we should keep